### PR TITLE
[build] Turn on -Werror on Linux and Mac platforms

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = #3335, #2927

Updated docs for compilation warnings:
https://deploy-preview-4928--docsite-preview.netlify.app/lang/articles/contribution/contributor_guide#compilation-warnings

### CODE REVIEW

**Review by ailzhang — APPROVED** [review]
Thanks for enabling this!!

File: docs/lang/articles/contribution/contributor_guide.md [hunk 1]
```diff
@@ -293,6 +293,61 @@ Here, we do not want to repeat some best practices summarized in the following G
   - [Code Health: Respectful Reviews == Useful Reviews](https://testing.googleblog.com/2019/11/code-health-respectful-reviews-useful.html)
   - [How to have your PR merged quickly](https://testing.googleblog.com/2017/06/code-health-too-many-comments-on-your.html)
 
+## Compilation Warnings
+Taichi enforces warning-free codes by turning on `-Werror` (treat warning as error) by default. It is highly recommended to resolve a warning as soon as it raises.
+
+On the other hand, real world issues could be way complexer than what the compiler expected. So we prepared the following HOWTOs to help resolve some common problems. You are also more than welcome to open up an issue or consult the reviewer inplace for further discussions.
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply docs/lang/articles/contribution/contributor_guide.md]: Done, thanks!

File: docs/lang/articles/contribution/contributor_guide.md [hunk 2]
```diff
@@ -293,6 +293,61 @@ Here, we do not want to repeat some best practices summarized in the following G
...
+```
+# Treat files under "external/Vulkan-Headers/include" as system headers, warnings of which will be muted.
+include_directories(SYSTEM external/Vulkan-Headers/include)
+
+# Treat files under "external/VulkanMemoryAllocator/include" as system headers for target "${CORE_LIBRARY_NAME}"
+target_include_directories(${CORE_LIBRARY_NAME} SYSTEM PRIVATE external/VulkanMemoryAllocator/include)
+```
+
+### How to deal with warnings raised when compiling third-party libraries or targets
+Ideally, third-party submodules should be built completely independent of Taichi project except for the topological dependency. Unfortunately, due to the design of CMake system, CMake Variables from Taichi and its submodules could get massed up in certain circumstances. Refer to the following two steps to mute warnings from third-party targets.
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply docs/lang/articles/contribution/contributor_guide.md]: Done, thanks!

File: docs/lang/articles/contribution/contributor_guide.md [hunk 3]
```diff
@@ -293,6 +293,61 @@ Here, we do not want to repeat some best practices summarized in the following G
...
+# Nullability seems to be a clang-specific feature, thus we disable this warning.
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-nullability-completeness ")
+
+# [Global] By evaluating "constexpr", compiler throws a warning for functions known to be dead at compile time.
+# However, some of these "constexpr" are debug flags and will be manually enabled upon debuging.
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unneeded-internal-declaration ")
+```
+
+### How to mute warnings for specific lines of codes
+Wrap the lines of interest with the following two macros, warnings will be ignored for the codes in between.
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply docs/lang/articles/contribution/contributor_guide.md:336]: Marked this section as "Not Recommended" and explained potential issues with using "#pragma clang diagnostics".

**Pull request merged**